### PR TITLE
[13.0][IMP] storage_backend_ftp: Implement Explicit FTP over TLS

### PR DESCRIPTION
**Pull request by JasminSForgeFlow**
@ForgeFlow

### CODE REVIEW

**Comment by JordiMForgeFlow** [comment]
LGTM 👍🏼 @JasminSForgeFlow could you fix pre-commit?

**Comment by JasminSForgeFlow** [comment]
> LGTM 👍🏼 @JasminSForgeFlow could you fix pre-commit?

That's I think cache issue we can try after 2 - 3 days again

**Comment by LoisRForgeFlow** [comment]
Rebased.

**Review by LoisRForgeFlow — APPROVED** [review]
LGTM

**Review by JordiMForgeFlow — APPROVED** [review]
Tested, LGTM 👍

File: storage_backend_ftp/components/ftp_adapter.py [hunk 1]
```diff
@@ -64,27 +64,33 @@ def sock(self, value):
...
+            ftp = ftplib.FTP()
+        elif backend.ftp_encryption == "tls":
+            ftp = ImplicitFTPTLS()
+            # Due to a bug into between ftplib and ssl, this part (about ssl) might not work!
+            # https://bugs.python.org/issue31727
+            security = FTP_SECURITY_TO_PROTOCOL.get(backend.ftp_security, None)
+            if isinstance(security, str):
+                raise UserError(security)
+        else:
+            raise NotImplementedError()
```
Comment (hparfr) [inline storage_backend_ftp/components/ftp_adapter.py]: dead code

**Review by hparfr — CHANGES_REQUESTED** [review]
Can you please refactor the new code with the old one. It's mostly the same actions: instantiate, login, (is it expected to not call connect()?), prot_p(), yield.

File: storage_backend_ftp/components/ftp_adapter.py (hunk 1 above)
Comment (JasminSForgeFlow) [reply storage_backend_ftp/components/ftp_adapter.py]: Hi,

Can you please suggest what change we can do here?
 
Thanks

**Comment by JasminSForgeFlow** [comment]
> Can you please refactor the new code with the old one. It's mostly the same actions: instantiate, login, (is it expected to not call connect()?), prot_p(), yield.

Hi,

Regarding this actually, we are not calling connect() method that's why we keep it a separate portion.

Thanks

**Review by MiquelRForgeFlow — APPROVED** [review]

File: storage_backend_ftp/__manifest__.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 {
     "name": "Storage Backend FTP",
     "summary": "Implement FTP Storage",
-    "version": "13.0.1.1.0",
+    "version": "13.0.1.1.1",
```
Comment (MiquelRForgeFlow) [inline storage_backend_ftp/__manifest__.py]: don't do this, this is done by merge bot

Comment (JasminSForgeFlow) [reply storage_backend_ftp/__manifest__.py]: Done, Thanks

File: storage_backend_ftp/components/ftp_adapter.py (hunk 1 above)
Comment (rousseldenis) [reply storage_backend_ftp/components/ftp_adapter.py]: @hparfr

Comment (hparfr) [reply storage_backend_ftp/components/ftp_adapter.py]:
```py 
if something in [a, b]:
  if something == a:
  elif something == b:
  else:
    dead code <-- here and line above (else)!
```
Just remove else and raise()

**Comment by hparfr** [comment]
> Regarding this actually, we are not calling connect() method that's why we keep it a separate portion.

Not tested but you may write something like that
(may be client.ssl_version should be used by tls_explicit)

```py
@contextmanager
def ftp(backend):
    security = None
    prot_p = False
    if backend.ftp_encryption in ["ftp", "tls", "tls_explicit"]:
        if backend.ftp_encryption == "ftp":
            ftp = ftplib.FTP()
        elif backend.ftp_encryption == "tls":
            ftp = ImplicitFTPTLS()
            # Due to a bug into between ftplib and ssl, this part (about ssl) might not work!
            # https://bugs.python.org/issue31727
            security = FTP_SECURITY_TO_PROTOCOL.get(backend.ftp_security, None)
            prot_p = True
            if isinstance(security, str):
                raise UserError(security)
        elif backend.ftp_encryption == "tls_explicit":
            ftp = ftplib.FTP_TLS()
            prot_p = True
        with ftp as client:
            if security:
                client.ssl_version = security
            client.connect(host=backend.ftp_server, port=backend.ftp_port)
            client.login(backend.ftp_login, backend.ftp_password)
            if prot_p:
                client.prot_p()
            if backend.ftp_passive:
                client.set_pasv(True)
            yield client
``

**Comment by JasminSForgeFlow** [comment]
> > Regarding this actually, we are not calling connect() method that's why we keep it a separate portion.
> 
> Not tested but you may write something like that (may be client.ssl_version should be used by tls_explicit)
> 
> ```python
> @contextmanager
> def ftp(backend):
>     security = None
>     prot_p = False
>     if backend.ftp_encryption in ["ftp", "tls", "tls_explicit"]:
>         if backend.ftp_encryption == "ftp":
>             ftp = ftplib.FTP()
>         elif backend.ftp_encryption == "tls":
>             ftp = ImplicitFTPTLS()
>             # Due to a bug into between ftplib and ssl, this part (about ssl) might not work!
>             # https://bugs.python.org/issue31727
>             security = FTP_SECURITY_TO_PROTOCOL.get(backend.ftp_security, None)
>             prot_p = True
>             if isinstance(security, str):
>                 raise UserError(security)
>         elif backend.ftp_encryption == "tls_explicit":
>             ftp = ftplib.FTP_TLS()
>             prot_p = True
>         with ftp as client:
>             if security:
>                 client.ssl_version = security
>             client.connect(host=backend.ftp_server, port=backend.ftp_port)
>             client.login(backend.ftp_login, backend.ftp_password)
>             if prot_p:
>                 client.prot_p()
>             if backend.ftp_passive:
>                 client.set_pasv(True)
>             yield client
> `` 
> ```

Yes, it should work.
updated same.

Thanks

**Review by hparfr — COMMENTED** [review]
/ocabot merge minor

**Review by hparfr — APPROVED** [review]

**Comment by hparfr** [comment]
The PR in v14 and v15 should include the last commit.

**Pull request merged**